packages/compass-query-bar/src/components/query-bar/query-bar.tsx

```diff
@@ -49,74 +61,260 @@ const openQueryHistoryStyles = cx(
-            htmlFor="open-query-history"
+  const onReset = useCallback(
+    (evt: React.MouseEvent) => {
+      // Prevent form submission.
+      evt.preventDefault();
+
+      _onReset();
+    },
+    [_onReset]
+  );
+
+  const toggleQueryHistory = useCallback(
+    (evt: React.MouseEvent) => {
+      // Prevent form submission.
+      evt.preventDefault();
+
+      _toggleQueryHistory();
+    },
+    [_toggleQueryHistory]
+  );
+
+  const onApply = useCallback(() => {
+    if (isQueryValid) {
+      _onApply();
+    }
+  }, [_onApply, isQueryValid]);
+
+  const renderQueryOption = useCallback(
+    (queryOption: QueryOption) => {
+      const hasError = !queryOptionProps[`${queryOption}Valid`];
+
+      const placeholder =
+        queryOptionProps[`${queryOption}Placeholder`] ||
+        OPTION_DEFINITION[queryOption].placeholder;
+
+      return (
+        <QueryOptionComponent
+          autoPopulated={autoPopulated}
+          hasError={hasError}
+          key={`query-option-${queryOption}`}
+          queryOption={queryOption}
+          onChange={(value: string) => onChangeQueryOption(queryOption, value)}
+          onApply={onApply}
+          placeholder={placeholder}
+          refreshEditorAction={refreshEditorAction}
+          schemaFields={schemaFields}
+          serverVersion={serverVersion}
+          value={queryOptionProps[`${queryOption}String`]}
+        />
+      );
+    },
+    [
+      autoPopulated,
+      onApply,
+      onChangeQueryOption,
+      refreshEditorAction,
+      schemaFields,
+      serverVersion,
+      queryOptionProps,
+    ]
+  );
+
+  const renderQueryOptionRow = useCallback(
+    (queryOption: string | string[], key: number) => (
+      <div
+        className={cx(rowStyles, key === 0 && firstQueryOptionsRowStyles)}
+        key={key}
+      >
+        {typeof queryOption === 'string'
+          ? renderQueryOption(queryOption as QueryOption)
+          : queryOption.map((optionName: string) =>
+              renderQueryOption(optionName as QueryOption)
+            )}
+        {key === layout.length - 1 && (
+          <Link
+            className={queryDocsLinkStyles}
+            href="https://docs.mongodb.com/compass/current/query/filter/"
+            target="_blank"
           >
-            Query
-          </Label>
+            Learn more
+          </Link>
+        )}
+      </div>
+    ),
+    [renderQueryOption]
+  );
+
+  const firstRowQueryOptions = useMemo(
+    () =>
+      layout.map((queryOption: string | string[], index: number) =>
+        index === 0 ? renderQueryOptionRow(queryOption, index) : null
+      ),
+    [layout, renderQueryOptionRow]
+  );
+
+  const additionalQueryOptions = useMemo(
+    () =>
+      layout.map((queryOption: string | string[], index: number) =>
+        isQueryOptionsExpanded && index > 0
+          ? renderQueryOptionRow(queryOption, index)
+          : null
+      ),
+    [isQueryOptionsExpanded, layout, renderQueryOptionRow]
+  );
```

Comment:
To make it more readable, would it make sense to just split the layout in two first and then just render it without using memo excessively? Something like:

```jsx
const [firstRow, ....rows] = layout;

return <>{renderRow(firstRow)}{isQueryOptionsExpanded && rows.map(row => renderRow(row))}</>
```

Reply:
Updated - added a `QueryBarRow` component let me know how it looks now. I think it will be a good bit more readable. I was definitely going a bit crazy with useMemo here